File: 2024/day19.rb
def count_valid(patterns, design)
  def check_valid(patterns, design, memo)
    return memo[design] if memo.include?(design)
    if design == ""
      memo[design] = 1
    else
      potential_patterns = patterns.filter { |p| design.start_with?(p) }
      if potential_patterns.empty?
        memo[design] = 0
      else
        memo[design] = potential_patterns.map { |p| check_valid(patterns, design[p.length..], memo) }.sum
      end
    end
  end
  check_valid(patterns, design, {})
end

lines = STDIN.readlines.map(&:chomp)
patterns = lines[0].split(', ')
designs = lines.drop(2)

num_valid = designs.map { |d| count_valid(patterns, d) }
puts num_valid.count { |d| d != 0 }
puts num_valid.sum

File: 2024/day19.ml
let remove_prefix p s = String.(sub s (length p) (length s - length p))
let remove_suffix p s = String.(sub s 0 (length s - length p))

module StringMap = Map.Make (struct
  type t = string

  let compare = compare
end)

let rec valid_design patterns design =
  let memo = StringMap.empty in
  let check_valid design =
    match StringMap.find_opt design memo with
    | Some v -> v
    | _ -> (
        if design = "" then true
        else
          match
            List.filter (fun p -> String.ends_with ~suffix:p design) patterns
          with
          | [] -> false
          | l ->
              List.map
                (fun p -> valid_design patterns (remove_suffix p design))
                l
              |> List.exists Fun.id)
  in
  check_valid design

let () =
  let lines = In_channel.input_lines stdin in
  let patterns, designs =
    match lines with
    | patternsStr :: _ :: designs ->
        let patterns =
          String.split_on_char ',' patternsStr |> List.map String.trim
        in
        (patterns, designs)
    | _ -> failwith "invalid input"
  in
  List.map (valid_design patterns) designs
  |> List.filter Fun.id |> List.length |> print_int
